### ISSUE TITLE
Strange checkpoint behavior

### ISSUE THREAD

**Issue by LuoXin-s** [PyTorchLightning/pytorch-lightning#12240]
## 🐛 Bug
When I test on a simple MNIST example with `ModelCheckpoint(monitor='train/loss', mode='min', save_top_k=1, save_last: true)`,  it behave correct. But when I switched from the demo to my project, there was a problem; since my dataset is large, it takes almost 12 hours for one epoch; I expected `ModelCheckpoint` to save the checkpoint per epoch like the demo. But after one day last, I found no checkpoint on my disk.

Further, I did the following test:
1: I use only the 0.1% percent dataset; the problem also exists.
2: I set `max_epochs=10`; when training is finished, the checkpoint dir has a "last.ckpt" checkpoint.
3: I set `every_n_train_steps=10`, then I can get checkpoint every 10 iteration correctly.

To summarize, when I set `every_n_train_steps,` I can get checkpoint every n step immediately. But when I set nothing(it equal to set 'every_n_epochs=1' as described in docs.

### Environment
PyTorch version: 1.10.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 16.04.7 LTS (x86_64)
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.23

Python version: 3.8.11 (default, Aug  3 2021, 15:09:35)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-3.10.0-1127.8.2.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080 Ti
Nvidia driver version: 470.74
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] pytorch-lightning==1.5.5
[pip3] torch==1.10.0
[pip3] torchaudio==0.10.0
[pip3] torchinfo==1.5.3
[pip3] torchmetrics==0.6.1
[pip3] torchvision==0.11.1
[conda] numpy                     1.19.5                   pypi_0    pypi
[conda] pytorch-lightning         1.5.5                    pypi_0    pypi
[conda] torch                     1.10.0                   pypi_0    pypi
[conda] torchaudio                0.10.0                   pypi_0    pypi
[conda] torchinfo                 1.5.3                    pypi_0    pypi
[conda] torchmetrics              0.6.1                    pypi_0    pypi
[conda] torchvision               0.11.1                   pypi_0    pypi

cc @carmocca @awaelchli @ninginthecloud @jjenniferdai @rohitgr7
